unlock_indicator.c

```diff
@@ -91,6 +94,99 @@ static double scaling_factor(void) {
 }
 
+/*
+ * Returns the cairo_font_face_t for sans-serif
+ * if this thing returns NULL, then we're somehow on a system without a
+ * sans-serif font, or a system without fontconfig (but with cairo somehow)
+ * Reference counts and caches the font-face, so we don't have memory leaks
+ * (and so that we don't have to go through all the parsing each time)
+ */
+
+static cairo_font_face_t *get_font_face() {
+    if (sans_serif) {
+        return sans_serif;
+    }
+
+    FcResult result;
+    /*
+     * Loads the default config.
+     * On successive calls, does no work and just returns true.
+     */
+    if (!FcInit()) {
+        DEBUG("Fontconfig init failed. No text will be shown.\n");
+        return NULL;
+    }
+    /*
+     * converts a font face name to a pattern for that face name
+     */
+    FcPattern *pattern = FcNameParse((const unsigned char *)"sans-serif");
+    if (!pattern) {
+        DEBUG("no sans-serif font available\n");
+        return NULL;
+    }
+
+    /*
+     * Gets the default font for our pattern. (Gets the default sans-serif font face)
+     * Without these two calls, the FcFontMatch call will fail due to FcConfigGetCurrent()
+     * not giving it a valid/useful config.
+     */
+    FcDefaultSubstitute(pattern);
+    if (!FcConfigSubstitute(FcConfigGetCurrent(), pattern, FcMatchPattern)) {
+        DEBUG("config sub failed?\n");
+        return NULL;
+    }
+    /*
+     * Looks up the font pattern and does some internal RenderPrepare work,
+     * then returns the resulting pattern that's ready for rendering.
+     */
+    FcPattern *pattern_ready = FcFontMatch(FcConfigGetCurrent(), pattern, &result);
+    FcPatternDestroy(pattern);
+    pattern = NULL;
+    if (!pattern_ready) {
+        DEBUG("no sans-serif font available\n");
+        return NULL;
+    }
+
+    /*
+     * Passes the given pattern into cairo, which loads it into a cairo freetype font face.
+     * Increment its reference count and cache it.
+     */
+    cairo_font_face_t *face = cairo_ft_font_face_create_for_pattern(pattern_ready);
+    FcPatternDestroy(pattern_ready);
+    sans_serif = cairo_font_face_reference(face);
+    return face;
+}
+
+/*
+ * Draws the given text onto the cairo context, using glyphs
+ */
+
+static void draw_text(cairo_t *ctx, const char *text, double offset) {
+    cairo_status_t status;
+    cairo_text_extents_t extents;
+    cairo_glyph_t *glyphs = NULL;
+    int num_glyphs;
+    double x, y;
+
+    cairo_text_extents(ctx, text, &extents);
+    x = BUTTON_CENTER - ((extents.width / 2) + extents.x_bearing);
+    y = BUTTON_CENTER - ((extents.height / 2) + extents.y_bearing) + offset;
+
+    status = cairo_scaled_font_text_to_glyphs(cairo_get_scaled_font(ctx),
+                                              x, y,
+                                              text, -1,
+                                              &glyphs, &num_glyphs,
+                                              NULL, NULL, NULL);
```

Comment:
I don't know why `cairo_show_text` is marked as toy API, but `cairo_scaled_font_text_to_glyphs` is not, but this code is (basically) doing what `cairo_show_text` would do (well, `cairo_show_text` uses and updates the current point instead of having an extra parameter for that, but that doesn't really matter, I think). So, I would suggest to keep using `cairo_show_text` here.